R/pseudocolor_plot.R

```diff
@@ -0,0 +1,346 @@
+#' @param pt.size Size of the points.
+#' @param alpha Transparency of the points.
+#' @param layer Name of layer to plot.
+#' @param grid_n Number of grid points to calculate the density.
+#' @param ... Additional arguments to pass to 'MASS::kde2d'.
+#'
+#' @return A ggplot object.
+#'
+#' @import rlang
+#'
+#' @examples
+#'
+#' library(pixelatorR)
+#' library(Seurat)
+#'
+#' # A mock-up Seurat Object
+#' object <-
+#'  CreateSeuratObject(counts = matrix(c(rpois(100000, 40),
+#'                                       rpois(100000, 5))[sample(1:200000, 200000)],
+#'                                     nrow = 100, ncol = 2000,
+#'                                     dimnames = list(paste0("Feature", 1:100),
+#'                                                     paste0("Cell", 1:2000))))
+#'
+#' object <-
+#'   AddMetaData(object,
+#'               metadata = data.frame(sample = rep(c("A", "B"), each = 1000),
+#'                                     sample_type = rep(c("Unstimulated", "Stimulated"),
+#'                                                       each = 500, times = 2),
+#'                                     row.names = paste0("Cell", 1:2000)))
+#'
+#' plot_gate <-
+#'   data.frame(xmin = c(70, 75),
+#'              xmax = c(150, 155),
+#'              ymin = c(50, 50),
+#'              ymax = c(150, 150),
+#'              sample = c("A", "B"))
+#'
+#'
+#' pseudocolor_plot(object,
+#'                  marker1 = "Feature1",
+#'                  marker2 = "Feature2",
+#'                  facet_vars = "sample",
+#'                  plot_gate = plot_gate,
+#'                  layer = "counts")
+#'
+#' @export
+#'
+pseudocolor_plot <- function(
+    object,
+    marker1,
+    marker2,
+    facet_vars = NULL,
+    plot_gate = NULL,
+    scale_density = TRUE,
+    margin_density = FALSE,
+    coord.fixed = T,
+    pt.size = 1,
+    alpha = 1,
+    layer = "data",
+    grid_n = 500,
+    ...
+) {
+
+  # Validate input parameters
+  stopifnot(
+    "'facet_vars' must either be NULL or be a character vector with 1 or 2 elements" =
+      is.null(facet_vars) | length(facet_vars) %in% 1:2,
+    "Variables in 'facet_vars' must be available in the object" =
+      is.null(facet_vars) | all(facet_vars %in% colnames(object[[]])),
+
+    "'marker1' must be a character of length 1" =
+      is.character(marker1),
+    "'marker2' must be a character of length 1" =
+      is.character(marker2),
+    "'marker1' must be available in the object" =
+      marker1 %in% rownames(object),
+    "'marker2' must be available in the object" =
+      marker2 %in% rownames(object),
+
+    "'grid_n' must be a numeric of length 1" =
+      is.numeric(grid_n),
+    "'scale_density' must be either TRUE or FALSE" =
+      is.logical(scale_density),
+    "'margin_density' must be either TRUE or FALSE" =
+      is.logical(margin_density),
+
+    "'pt.size' must be a numeric of length 1" =
```

Comment:
We have mostly used snake case for variable names, i.e. `pt_size`. Would that be fine to use here as well?